client/utils/guidance.py

```diff
@@ -0,0 +1,210 @@
+    'none': lambda info, strMisU: None, 
+    'ki': lambda info, strMisU: info['ki'],
+    'misU2Msg': lambda info, strMisU: info['dictMisU2Msg'].get(strMisU)
+}
+  
+class Guidance():
+    def __init__(self,current_working_dir):
+        """
+        Initializing everything we need to the default values. If we catch
+        an error when opening the JSON file, we flagged it as error.
+        """
+        self.tg_id = -1
+        self.misU_count = {}
+        self.current_working_dir = current_working_dir
+        try:
+            with open(current_working_dir + OK_GUIDANCE,"r") as f:
+                self.guidance_json = json.load(f)
+            self.error = False
+        except:
+            self.error = True
+
+
+    def guidance_msg(self,unique_id, input_lines,access_token,hash_key,guidance_flag):
+        """ 
+        Based on the student's answer (input_lines), we grab each associated
+        message if its corresponding misunderstanding's count is above the threshold
+        """    
+        if self.error:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        shorten_unique_id = assess_id_util.canon(unique_id)
+        # Try to get th info dictionary for this question
+        wa_2_dict_info = self.guidance_json['dictAssessId2WA2DictInfo'].get(shorten_unique_id)
+
+        if not wa_2_dict_info: # get returns None if wa_2_dict_info doesn't have shorten_unique_id in dictionary
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        dict_info = wa_2_dict_info.get(repr(input_lines))
+
+        # If this wrong answer is not in the JSON file, display default message
+        if not dict_info:            
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        self.set_tg(access_token,guidance_flag)
+        if self.tg_id == TG_ERROR_VALUE: 
+            # If self.tg_id == -1, some errors happen when trying to access server
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        lambda_string_key = self.guidance_json['dictTg2Func'].get(str(self.tg_id))
+
+        if not lambda_string_key:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        lambda_info_misu = lambda_string_key_to_func.get(lambda_string_key)
+        if not lambda_info_misu:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        lst_mis_u = dict_info.get('lstMisU')
+
+        # No list of misunderstandings for this wrong answer, default message
+        if not lst_mis_u:
+            print (DEFAULT_MSG)
+            return DEFAULT_RETURN_VALUE
+
+        self.misU_count = self.update_misUcounts(hash_key, lst_mis_u, repr(input_lines),shorten_unique_id)
+
+        threshold = self.guidance_json['wrongAnsThresh']
+
+        self.msg_id_set = set()
+
+        # If the count is higher than the threshold, we need to display the message
+        for mis_u in self.misU_count:
```

Comment:
I would have thought `misU_count` was an integer because of `count`. Perhaps rename to `misU_counts`.